src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php

```diff
@@ -82,13 +86,38 @@ public function process(ContainerBuilder $container)
                         $event['method'] = '__invoke';
                     }
                 }
+                $registeredMethods[$event['method']] = true;
 
                 $definition->addMethodCall('addListener', [$event['event'], [new ServiceClosureArgument(new Reference($id)), $event['method']], $priority]);
-
                 if (isset($this->hotPathEvents[$event['event']])) {
                     $container->getDefinition($id)->addTag($this->hotPathTagName);
                 }
             }
+
+            if ($isAutoregisterEnabled) {
+                if (null === ($class = $container->getDefinition($id)->getClass()) || null === ($r = $container->getReflectionClass($class, false))) {
+                    throw new InvalidArgumentException(sprintf('Service "%s" must define the "event" attribute on "%s" tags.', $id, $this->listenerTag));
+                }
+                $hasAutoregisteredMethod = false;
+                foreach ($r->getMethods(\ReflectionMethod::IS_PUBLIC) as $m) {
+                    if ($m->isConstructor() || $m->isDestructor() || $m->isStatic() || $m->isAbstract() || $m->isGenerator() || isset($registeredMethods[$m->getName()])) {
+                        continue;
+                    }
+                    $parameters = $m->getParameters();
+                    if ($m->getNumberOfRequiredParameters() < 1 || !$parameters[0]->hasType() || !\is_a(($eventName = $parameters[0]->getType()->getName()), Event::class, true)) {
+                        continue;
+                    }
+                    $hasAutoregisteredMethod = true;
+
+                    $definition->addMethodCall('addListener', [$eventName, [new ServiceClosureArgument(new Reference($id)), $m->getName()], 0]);
```

Comment:
maybe some lines to share with the existing code?

Reply:
refactored (fixed an hardcoded `'container.hot_path'` in eventSubscriber. Was it on purpose?